### ISSUE TITLE
Rescue master branch commits

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/silverstripe-framework#10350]
There's still some potentially useful work in the master branches. Those were meant to be to bu sue for CMS 5, however they are horribly out of date. There might be some useful commits there that we want to fold back into the new major release branches.

## Acceptance criteria
- [x] Commits that are unique to the master branches and are useful are cherry-picked tot he new major branches.
- [x] Commits don't make sense with work that got completed after they got merged into master get left behind.
- [x] Commits that wold introduce major upgrade pains are left behind.
- [x] Breaking changes are noted in the appropriate documentation
- [x] Create follow task to delete old master branches to give people a chance to rescue any commits they care about.

Depends on #10349

## Notes
- There was a spike for this which may be useful: https://github.com/silverstripeltd/product-issues/issues/539

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/10450
  - https://github.com/silverstripe/silverstripe-postgresql/pull/135
- https://github.com/silverstripe/silverstripe-framework/pull/10454
- https://github.com/silverstripe/silverstripe-framework/pull/10456
- https://github.com/silverstripe/silverstripe-framework/pull/10457
  - https://github.com/silverstripe/silverstripe-cms/pull/2764
  - https://github.com/silverstripe/silverstripe-versioned/pull/371
- https://github.com/silverstripe/silverstripe-framework/pull/10458
- https://github.com/silverstripe/silverstripe-framework/pull/10459
- https://github.com/silverstripe/silverstripe-framework/pull/10460
- https://github.com/silverstripe/silverstripe-framework/pull/10461
- https://github.com/silverstripe/silverstripe-framework/pull/10462
- https://github.com/silverstripe/silverstripe-framework/pull/10463
- https://github.com/silverstripe/silverstripe-cms/pull/2765
- https://github.com/silverstripe/silverstripe-cms/pull/2766
- https://github.com/silverstripe/silverstripe-cms/pull/2768
- https://github.com/silverstripe/silverstripe-cms/pull/2769
- https://github.com/silverstripe/silverstripe-admin/pull/1346
- https://github.com/silverstripe/silverstripe-admin/pull/1349
- https://github.com/silverstripe/silverstripe-assets/pull/507
- https://github.com/silverstripe/silverstripe-assets/pull/508
- https://github.com/silverstripe/silverstripe-versioned-admin/pull/249
- https://github.com/silverstripe/silverstripe-versioned/pull/372
- https://github.com/silverstripe/silverstripe-reports/pull/154
- https://github.com/silverstripe/silverstripe-elemental/pull/1002
- https://github.com/silverstripe/developer-docs/pull/70
- https://github.com/silverstripe/silverstripe-framework/pull/10484 (again)
    - https://github.com/silverstripe/silverstripe-postgresql/pull/138 (again)
    - https://github.com/silverstripe/silverstripe-postgresql/pull/139
 - https://github.com/silverstripe/silverstripe-framework/pull/10497
 - https://github.com/silverstripe/silverstripe-framework/pull/10507
 - https://github.com/silverstripe/developer-docs/pull/81

## PRs for broken builds
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1290
- https://github.com/silverstripe/silverstripe-frameworktest/pull/121

**Comment by GuySartorelli** [comment]
https://github.com/silverstripe/silverstripe-sqlite3/pull/44 will need to be rescued once a new major release branch is created for that module.

**Comment by GuySartorelli** [comment]
@silverstripe/core-team 

I identified the following master branch PRs as potentially controversial - they may introduce more upgrade pain than we want to introduce for this major release:
- https://github.com/silverstripe/silverstripe-framework/pull/8152
  - Accompanied by https://github.com/silverstripe/silverstripe-framework/pull/8456
- https://github.com/silverstripe/silverstripe-framework/pull/8148

Do any of you have opinions on whether we should rescue those PRs for inclusion in CMS 5?

**Comment by GuySartorelli** [comment]
Follow up card to remove master branches has been created: https://github.com/silverstripe/silverstripe-framework/issues/10472

**Comment by lozcalver** [comment]
#8152 and #8456 seem relatively low risk to me - the only BC issues I can think of are:

- Someone passing `true`, `false`, `null` or a number and expecting strings rather than scalars/null. I think it’s unlikely, and PHP’s loose typing makes it less likely to be an issue
- Someone typehinting strings for method args

Both cases would be highlighted pretty clearly by errors when the templates run, and would be easy to fix. If anyone needs to preserve the current behaviour, they can quote any values they’ve got hardcoded in templates (it only applies to `$Foo(123)`, not `$Foo($SomethingReturningAnInteger)`)

#8148 is the more controversial one because it likely won’t error - it’ll either stop outputting data, or output data from the wrong “scope”. The problem with that one is that there’s no way to fix #4015 without causing that upgrade pain - so either we suck it up at some point or never fix it 😕

**Comment by emteknetnz** [comment]
I'm happy to have all three merged.

Applications currently passing `false` or `null` from a template would be currently working by accident because there is broken logic in the controller i.e. `if ($myTemplateVar) ... ` - string `"false"` is truthy. So the fact it breaks on upgrade is actually what's supposed to happen.

For the `$Up.Title`, it's very easy to provide clear upgrade instructions here, simply do a project string find and replace for possible template strings such ` $Up.` (with a leading space) or `($Up.` in template files.

**Comment by emteknetnz** [comment]
All linked PRs have been merged

**Comment by GuySartorelli** [comment]
Reopening - there is still the discussion about the potentially controversial PRs which have not yet been rescued.

**Comment by maxime-rainville** [comment]
I also don't think any of those changes are likely to cause a great deal of upgrade headaches. I'm sure it will trip somebody somewhere, but that seems reasonable.

On #8148, it's not clear to me the new behaviour is obviously better than the old one. I don't really have strong feeling either way. Looking at the matching https://github.com/silverstripe/silverstripe-framework/issues/4015 issue, I see a mix of view there as well. 

My instinct would be to
- recover 
  - #8152
  - #8456
- let #8148 go

**Comment by GuySartorelli** [comment]
I agree with Steve that it's pretty easy to call out how to upgrade the scope change (and for a major release people are much more likely to actually look at the changelog to see what they may need to change), and the majority consensus seems to be that it's a worthwhile change - I'll rescue all three PRs.

**Comment by GuySartorelli** [comment]
Blocked pending alpha1 being tagged.
